### PR TITLE
[BREAKING] Revert if calldata has wrong size

### PR DESCRIPTION
**Pull request by leonardoalt**
Implements https://github.com/ethereum/solidity/issues/2109 and https://github.com/ethereum/solidity/issues/3979.

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ Breaking Changes:
  * Type Checker: Disallow arithmetic operations for Boolean variables.
  * Disallow trailing dots that are not followed by a number.
  * Remove assembly instructions ``sha3`` and ``suicide``
+ * Code generator: revert if calldata has wrong size.
```
Comment (chriseth) [inline Changelog.md]: We only revert if it is too short. Perhaps also mention that the revert happens inside the ABI decoder and this also applies to `abi.decode()` (where this has been the behaviour already before).

**Comment by leonardoalt** [comment]
Tests fixed and rebased.

**Comment by axic** [comment]
It breaks this zeppeling test:
> Contract: ECRecovery recover should fail when a wrong hash is sent:

File: Changelog.md [hunk 2]
```diff
@@ -19,6 +19,7 @@ Breaking Changes:
  * Type Checker: Disallow arithmetic operations for boolean variables.
  * Type Checker: Disallow conversions between ``bytesX`` and ``uintY`` of different size.
  * Remove obsolete ``std`` directory from the Solidity repository. This means accessing ``https://github.com/ethereum/soldity/blob/develop/std/*.sol`` (or ``https://github.com/ethereum/solidity/std/*.sol`` in Remix) will not be possible.
+ * Code generator: revert if calldata is too short or points out of bounds. This is done inside the ``ABI decoder`` and therefore also applies to ``abi.decode()``.
```
Comment (axic) [inline Changelog.md]: While this should be probably self explanatory, maybe adding "revert at runtime if calldata" is useful for those quickly grepping the changelog.

**Comment by axic** [comment]
I think it looks good otherwise.

**Review by axic — APPROVED** [review]

**Pull request merged**